src/Updater.cpp

```diff
@@ -0,0 +1,108 @@
+/*
+ * This file is part of Adblock Plus <https://adblockplus.org/>,
+ * Copyright (C) 2006-present eyeo GmbH
+ *
+ * Adblock Plus is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * Adblock Plus is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Adblock Plus.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <AdblockPlus.h>
+#include "Utils.h"
+#include <cassert>
+
+using namespace AdblockPlus;
+
+namespace {
+    static std::string updaterJsFiles[] = {
+      "compat.js",
+      "info.js",
+      "prefs.js",
+      "utils.js",
+      "events.js",
+      "coreUtils.js",
+      "common.js",
+      "downloader.js",
+      "updater.js",
+      "apiUpdater.js"
+    };
+}
+
+Updater::Updater(const JsEnginePtr& jsEngine, const JsEngine::EvaluateCallback& evaluateCallback)
+  : jsEngine(jsEngine), updateCheckId(0)
+{
+  // Hack to enable downloads from JS (see compat.js)
+  jsEngine->SetEventCallback("_isSubscriptionDownloadAllowed", [this](JsValueList&& params)
+  {
+    // param[0] - nullable string Prefs.allowed_connection_type
+    // param[1] - function(Boolean)
+    bool areArgumentsValid = params.size() == 2 && (params[0].IsNull() || params[0].IsString()) && params[1].IsFunction();
+    assert(areArgumentsValid && "Invalid argument: there should be two args and the second one should be a function");
+    if (!areArgumentsValid)
+      return;
+    params[1].Call(this->jsEngine->NewValue(true));
+  });
+
+  // Set empty preconfigured prefs
+  auto preconfiguredPrefsObject = jsEngine->NewObject();
+  jsEngine->SetGlobalProperty("_preconfiguredPrefs", preconfiguredPrefsObject);
```

Comment:
This code is actually unrelated, however, since prefs.js is required and prefs.js expects that object, I would propose to change prefs.js. The latter should rather check whether `_preconfiguredPrefs` exists. Additionally, it can happen that someone firstly instantiated `Updater` and then is instantiating `FilterEngine`, but prefs are already initialized. IMO we may not fully address it here but fix it later, under another ticket.

Reply:
I'll remove this code from the Updater class and add a check to prefs.js to iterate through _preconfiguredPrefs only if this object exists. Is that enough for the scope of this task? Later on _preconfiguredPrefs could be split so FilterEngine and Updater would use dedicated objects.

Reply:
It's absolutely enough. I think later we should rather extract Prefs the similar way as Updater and make the former as a dependency for FilterEngine and Updater.